File: 剑指Offer/07-重建二叉树/c-c++/7_1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
    {
        if (preorder.size() == 0 || inorder.size() == 0)
            return NULL;

        // 递归开始判断
        return buildTreeCore(preorder, inorder, 0, preorder.size() - 1, 0, inorder.size() - 1);
    }

    TreeNode *buildTreeCore(vector<int> &preorder, vector<int> &inorder, int preStart, int preEnd, int inStart, int inEnd)
    {
        // 构造当前节点
        TreeNode *root = new TreeNode(preorder[preStart]);
        int rootValue = root->val;

        // 如果当前是最后一个节点了, 直接返回
        if (preStart == preEnd)
        {
            if (inStart == inEnd && preorder[preStart] == inorder[inStart])
                return root;
        }

        // 在中序遍历中找到根节点的值
        int index = inStart;
        while (inorder[index] != rootValue)
            index++;

        // 得到左子树的长度
        int leftLength = index - inStart;

        // 递归遍历左子树
        if (leftLength > 0)
            root->left = buildTreeCore(preorder, inorder, preStart + 1, preStart + leftLength - 1, inStart, inStart + leftLength - 1);
        // 递归遍历右子树
        if (index < inEnd)
            root->right = buildTreeCore(preorder, inorder, preStart + 1 + leftLength, preEnd, index + 1, inEnd);

        return root;
    }
};

File: 剑指Offer/07-重建二叉树/go/7_1.go
package tree

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) <= 0 || len(inorder) <= 0 {
		return nil
	}

	return _buildTree(preorder, inorder, 0, len(preorder)-1, 0, len(inorder)-1)
}

// 递归实现
func _buildTree(preorder []int, inorder []int, preStart, preEnd, inStart, inEnd int) *TreeNode {
	rootValue := preorder[preStart]

	// 前序遍历的第一个节点就是根节点
	root := &TreeNode{
		Val: rootValue,
	}

	// 如果当前只有一个节点了
	if preStart == preEnd {
		if inStart == inEnd && preorder[preStart] == inorder[inStart] {
			return root
		}
	}

	// 在中序遍历中找到根节点的位置
	rootValueInInOrder := inStart
	for ; inorder[rootValueInInOrder] != rootValue && rootValueInInOrder <= inEnd; rootValueInInOrder++ {
	}

	// 左子节点的数量
	leftCount := rootValueInInOrder - inStart

	// 递归构建左子结点
	if leftCount > 0 {
		root.Left = _buildTree(preorder, inorder, preStart+1, preStart+leftCount, inStart, rootValueInInOrder-1)
	}

	// 递归构建右子节点
	if rootValueInInOrder < inEnd {
		root.Right = _buildTree(preorder, inorder, preStart+leftCount+1, preEnd, rootValueInInOrder+1, inEnd)
	}

	// 返回当前根节点
	return root
}
